File: figures/05-dynamic-programming/snippet-fib-memo.java
    private static long memoRec(int n, long[] memo) {
        if (memo[n] != Long.MIN_VALUE) return memo[n];
        memo[n] = Math.addExact(memoRec(n - 1, memo), memoRec(n - 2, memo));
        return memo[n];
    }

    public static long fibMemo(int n) {
        if (n < 0 || n > 92) throw new IllegalArgumentException("n must be 0..92");
        long[] memo = new long[n + 1];
        Arrays.fill(memo, Long.MIN_VALUE);
        memo[0] = 0;
        if (n >= 1) memo[1] = 1;
        return memoRec(n, memo);
    }

File: figures/05-dynamic-programming/snippet-fib-memo.c
static long long memo_rec(int n, long long *memo) {
    if (memo[n] != LLONG_MIN) return memo[n];
    memo[n] = memo_rec(n - 1, memo) + memo_rec(n - 2, memo);
    return memo[n];
}

static bool fib_memo(int n, long long *out) {
    if (!out || n < 0 || n > 92) return false;
    long long *memo = calloc((size_t)n + 1, sizeof(*memo));
    if (!memo) return false;
    for (int i = 0; i <= n; ++i) memo[i] = LLONG_MIN;
    memo[0] = 0;
    if (n >= 1) memo[1] = 1;
    *out = memo_rec(n, memo);
    free(memo); return true;
}
